external/openglcts/modules/common/glcPixelStorageModesTests.cpp

```diff
@@ -0,0 +1,2583 @@
+		(getNumberOfChannels(m_internalFormat) > 1 ? t_value : 0u),
+		(getNumberOfChannels(m_internalFormat) > 2 ? t_value : 0u),
+		255u,
+	};
+
+	gl.glUniform4uiv(gl.glGetUniformLocation(prog->getProgram(), "refcolour"), 1,
+					 refcol);
+}
+
+void TexImageUtils::setRGB10A2Refcolour (glu::CallLogWrapper gl, glw::GLdouble col)
+{
+	unsigned int max_channel_value = 1023u;
+
+	const glw::GLdouble d_max_channel_value = (glw::GLdouble)max_channel_value;
+	const glw::GLdouble d_value = (glw::GLdouble)d_max_channel_value * col;
+	unsigned int t_value = (unsigned int)d_value;
+
+	unsigned int refcol[4] =
+	{
+		(getNumberOfChannels(m_internalFormat) > 0 ? t_value : 0u),
+		(getNumberOfChannels(m_internalFormat) > 1 ? t_value : 0u),
+		(getNumberOfChannels(m_internalFormat) > 2 ? t_value : 0u),
+		255u,
+	};
+
+	gl.glUniform4uiv(gl.glGetUniformLocation(prog->getProgram(), "refcolour"), 1,
+					 refcol);
+}
+
+bool TexImageUtils::verify(tcu::Surface dst, tcu::Surface *errMask)
+{
+	*errMask = tcu::Surface (dst.getWidth(), dst.getHeight());
+	tcu::clear(errMask->getAccess(), tcu::Vec4(0.0f, 1.0f, 0.0f, 1.0f));
+	bool pass = true;
+
+	for (int y = 0; y < dst.getHeight(); y++)
+	{
+		for (int x = 0; x < dst.getWidth(); x++)
+		{
+			if (dst.getPixel(x, y) != tcu::RGBA::green())
+			{
+				pass = false;
+				errMask->setPixel(x, y, tcu::RGBA::red());
+			}
+		}
+	}
+
+	return pass;
+}
+
+class TexImage2DCase : public deqp::TestCase
+					 , public sglr::ContextWrapper
+					 , public TexImageUtils
+{
+public:
+	TexImage2DCase (deqp::Context& context, const char* name, const char* desc,
+					deUint32 internalFormat,
+					int rect_w, int rect_h,
+					int subrect_x0, int subrect_y0,
+					int subrect_w, int subrect_h,
+					glu::GLSLVersion glsl_version);
+	~TexImage2DCase (void);
+	IterateResult iterate (void);
+protected:
+	void generateSrcData();
+	void createTexture (void);
+	void createShader (void);
+	tcu::Surface renderToSurf (void);
+};
+
+TexImage2DCase::TexImage2DCase (deqp::Context& context,
+								const char* name,
+								const char* desc,
+								deUint32 internalFormat,
+								int rect_w,
+								int rect_h,
+								int subrect_x0,
+								int subrect_y0,
+								int subrect_w,
+								int subrect_h,
+								glu::GLSLVersion glsl_version)
+	: TestCase(context, name, desc)
+	, TexImageUtils(internalFormat,
+					rect_w, rect_h, 1,
+					subrect_x0, subrect_y0, 0,
+					subrect_w, subrect_h, 1,
+					glsl_version)
+{
+}
+
+TexImage2DCase::~TexImage2DCase(void)
+{
+}
+
+TexImage2DCase::IterateResult TexImage2DCase::iterate(void)
+{
+	glu::RenderContext& renderCtx = TestCase::m_context.getRenderContext();
+	tcu::TestLog& log = m_testCtx.getLog();
+	tcu::Surface dst, errMask;
+
+	bool pass = true;
+
+	sglr::GLContext gl_ctx (renderCtx,
+							log,
+							sglr::GLCONTEXT_LOG_CALLS,
+							tcu::IVec4(0, 0, m_subcuboid_w, m_subcuboid_h));
+
+	setContext((sglr::Context*)&gl_ctx);
+
+	generateSrcData();
+	createTexture();
```

Comment:
are the gl texture and shader objects being cleaned up after each iteration?

Reply:
Texture wasn't. I moved resource deallocation stuff to `cleanup()` functions.